src/objects.rs

```diff
@@ -80,33 +185,75 @@ impl ObjectData {
         parser: &mut impl Iterator<Item = XmlEventResult>,
         attrs: Vec<OwnedAttribute>,
         tilesets: Option<&[MapTilesetGid]>,
+        for_tileset: Option<Arc<Tileset>>,
+        base_path: &Path,
+        reader: &mut impl ResourceReader,
+        cache: &mut impl ResourceCache,
     ) -> Result<ObjectData> {
-        let ((id, tile, n, t, w, h, v, r), (x, y)) = get_attrs!(
+        let (mut id, mut tile, mut x, mut y, mut n, mut t, mut w, mut h, mut v, mut r, template) = get_attrs!(
             attrs,
             optionals: [
                 ("id", id, |v:String| v.parse().ok()),
                 ("gid", tile, |v:String| v.parse().ok()
-                                            .and_then(|bits| LayerTileData::from_bits(bits, tilesets?))),
+                                            .and_then(|bits| ObjectTileData::from_bits(bits, tilesets?, for_tileset.as_ref().cloned()))),
+                ("x", x, |v:String| v.parse().ok()),
+                ("y", y, |v:String| v.parse().ok()),
                 ("name", name, |v:String| v.parse().ok()),
                 ("type", obj_type, |v:String| v.parse().ok()),
                 ("width", width, |v:String| v.parse().ok()),
                 ("height", height, |v:String| v.parse().ok()),
                 ("visible", visible, |v:String| v.parse().ok().map(|x:i32| x == 1)),
                 ("rotation", rotation, |v:String| v.parse().ok()),
-            ],
-            required: [
-                ("x", x, |v:String| v.parse().ok()),
-                ("y", y, |v:String| v.parse().ok()),
-            ],
-            Error::MalformedAttributes("objects must have an x and a y number".to_string())
+                ("template", template, |v:String| v.parse().ok()),
+            ]
         );
+
+        // If the template attribute is there, we need to go fetch the template file
+        let template = template
+            .map(|template| {
+                let s: String = template;
+                let parent_dir = base_path.parent().ok_or(Error::PathIsNotFile)?;
+                let template_path = parent_dir.join(Path::new(&s));
+
+                // Check the cache to see if this template exists
+                let template = if let Some(templ) = cache.get_template(&template_path) {
+                    templ
+                } else {
+                    let template = Template::parse_template(&template_path, reader, cache)?;
+                    // Insert it into the cache
+                    cache.insert_template(&template_path, template.clone());
+                    template
+                };
+
+                // The template sets the default values for the object
+                let obj = &template.object;
+                x.get_or_insert(obj.x);
+                y.get_or_insert(obj.y);
```

Comment:
Actually a template instance never inherits the `x` or `y` from the template object (these attributes are also never saved for the template object), so this effectively just makes then 0, which is already the default.